src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs

```diff
@@ -401,6 +415,203 @@ static void PromoteCommonCommitCharactersOntoList(LSP.VSCompletionList completio
+                var originalDocument = document.WithText(sourceText.WithChanges(new TextChange(completionListSpan, string.Empty)));
+                // We don't have access to the original trigger, but we know the completion list is already present.
+                // It is safe to recompute with the invoked trigger as we will get all the items and filter down based on the current trigger.
+                var originalTrigger = new CompletionTrigger(CompletionTriggerKind.Invoke);
+                return await CalculateListAsync(request, originalDocument, completionListSpan.Start, originalTrigger, completionOptions, completionService, completionListCache, cancellationToken).ConfigureAwait(false);
+            }
+
+            return cachedList;
+
+            static (CompletionList, long)? GetCachedList(
+                (SourceText SourceText, long ResultId)? lastResultId,
+                SourceText textWithoutCompletionListSpan,
+                CompletionListCache completionListCache)
+            {
+                if (lastResultId.HasValue)
+                {
+                    if (textWithoutCompletionListSpan.ContentEquals(lastResultId.Value.SourceText))
+                    {
+                        var cachedList = completionListCache.GetCachedCompletionList(lastResultId.Value.ResultId);
+                        if (cachedList != null)
+                        {
+                            return (cachedList.CompletionList, lastResultId.Value.ResultId);
+                        }
+                    }
+                }
+
+                return null;
+            }
+        }
+
+        private static async Task<(CompletionList CompletionList, long ResultId)?> CalculateListAsync(
+            LSP.CompletionParams request,
+            Document document,
+            int position,
+            CompletionTrigger completionTrigger,
+            OptionSet completionOptions,
+            CompletionService completionService,
+            CompletionListCache completionListCache,
+            CancellationToken cancellationToken)
+        {
+            var completionList = await completionService.GetCompletionsAsync(document, position, completionTrigger, options: completionOptions, cancellationToken: cancellationToken).ConfigureAwait(false);
+            cancellationToken.ThrowIfCancellationRequested();
+            if (completionList == null || completionList.Items.IsEmpty)
+            {
+                return null;
+            }
+
+            // Cache the completion list so we can avoid recomputation in the resolve handler
+            var resultId = completionListCache.UpdateCache(request.TextDocument, completionList);
+
+            return (completionList, resultId);
+        }
+
+        private static (CompletionList CompletionList, bool IsIncomplete) FilterCompletionList(CompletionList completionList, TextSpan completionListSpan, CompletionTrigger completionTrigger, SourceText sourceText, Document document)
```

Comment:
nit: I think wrapping the arguments here could be helpful for those with smaller screens (there's a couple of similar places above that also seem to be a little long)